In [35]:
#import libraries and modules
import pandas as pd
import numpy as np
from sklearn.naive_bayes import MultinomialNB
from sklearn.feature_extraction import stop_words
from sklearn.feature_extraction.text import CountVectorizer
from collections import Counter
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.metrics import classification_report
from sklearn.pipeline import make_pipeline
from sklearn.base import TransformerMixin
class DenseTransformer(TransformerMixin):
    
    def fit(self, X, y=None, **fit_params):
        return self
    
    def transform(self, X, y=None, **fit_params):
        return X.todense()
from sklearn.preprocessing import Normalizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LogisticRegressionCV
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.svm import LinearSVC

from collections import Counter


In [36]:
!pwd

/Users/Eric/Documents/DataAnalyticsBootcamp/MY_WORK_AREA/UTAUS201901DATA3/Projects/FinalProject/repo_world_wine_web_MY_SANDBOX/FinalProject-WorldWineWebML/models


In [37]:
#import the data
data = pd.read_csv("final_wine_data_172k_test.csv")
data.head()

Unnamed: 0   country                                        description  \
0           1  Portugal  This is ripe and fruity, a wine that is smooth...   
1           2        US  Tart and snappy, the flavors of lime flesh and...   
2           3        US  Pineapple rind, lemon pith and orange blossom ...   
3           4        US  Much like the regular bottling from 2012, this...   
4           5     Spain  Blackberry and raspberry aromas show a typical...   

   price  points             variety               winery  
0   15.0      87      Portuguese Red  Quinta dos Avidagos  
1   14.0      87          Pinot Gris            Rainstorm  
2   13.0      87            Riesling           St. Julian  
3   65.0      87          Pinot Noir         Sweet Cheeks  
4   15.0      87  Tempranillo-Merlot               Tandem

In [17]:
data = data.filter(["country", "description"], axis = 1)
data.head()

country                                        description
0  Portugal  This is ripe and fruity, a wine that is smooth...
1        US  Tart and snappy, the flavors of lime flesh and...
2        US  Pineapple rind, lemon pith and orange blossom ...
3        US  Much like the regular bottling from 2012, this...
4     Spain  Blackberry and raspberry aromas show a typical...

In [18]:
#look at the unique countries
data['country'].unique()

array(['Portugal', 'US', 'Spain', 'Italy', 'France', 'Germany',
       'Argentina', 'Chile', 'Australia', 'Austria', 'South Africa',
       'New Zealand', 'Israel', 'Hungary', 'Greece', 'Romania', 'Mexico',
       'Canada', 'Turkey', 'Czech Republic', 'Slovenia', 'Luxembourg',
       'Croatia', 'Georgia', 'Uruguay', 'England', 'Lebanon', 'Serbia',
       'Brazil', 'Moldova', 'Morocco', 'Peru', 'India', 'Bulgaria',
       'Cyprus', 'Armenia', 'Switzerland', 'Bosnia and Herzegovina',
       'Ukraine', 'Slovakia', 'Macedonia', 'China', 'Kosovo', 'Global'],
      dtype=object)

In [19]:
#look at the country balance
data['country'].value_counts()

US                        70956
Italy                     31757
France                    26183
Spain                      8609
Portugal                   6913
Chile                      5284
Argentina                  4567
Austria                    3866
Australia                  3148
Germany                    2917
New Zealand                1935
South Africa               1582
Israel                      689
Greece                      566
Canada                      336
Uruguay                     228
Bulgaria                    202
Hungary                     187
Romania                     143
England                     123
Georgia                     113
Turkey                      107
Moldova                     100
Croatia                      85
Slovenia                     84
Brazil                       73
Mexico                       73
Lebanon                      35
Morocco                      34
Switzerland                  20
Peru                         17
Ukraine 

In [20]:
data['length'] = data['description'].apply(lambda x: len(x))
data.head()

country                                        description  length
0  Portugal  This is ripe and fruity, a wine that is smooth...     227
1        US  Tart and snappy, the flavors of lime flesh and...     186
2        US  Pineapple rind, lemon pith and orange blossom ...     199
3        US  Much like the regular bottling from 2012, this...     249
4     Spain  Blackberry and raspberry aromas show a typical...     261

In [21]:
data.dtypes


country        object
description    object
length          int64
dtype: object

In [22]:
X=data['description']
y=data['country']

X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=42)

In [ ]:
### Linear Regression   --   maybe not approp --  dnf

## Modeling

### Categorical Naive Bayes

In [23]:
%%time

naive_bayes = make_pipeline(
    CountVectorizer(stop_words='english', binary=True),
    MultinomialNB()
)

naive_bayes.fit(X_train, y_train)

print(f'Accuracy: {naive_bayes.score(X_test, y_test)} \n')
print(classification_report(y_test, naive_bayes.predict(X_test)))

Accuracy: 0.8246813983397638 



/Users/Eric/anaconda3/lib/python3.7/site-packages/sklearn/metrics/classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


                        precision    recall  f1-score   support

             Argentina       0.70      0.35      0.47       954
               Armenia       0.00      0.00      0.00         3
             Australia       0.66      0.52      0.58       610
               Austria       0.51      0.74      0.60       760
Bosnia and Herzegovina       0.00      0.00      0.00         2
                Brazil       0.00      0.00      0.00        12
              Bulgaria       0.00      0.00      0.00        44
                Canada       1.00      0.02      0.03        63
                 Chile       0.62      0.47      0.54       996
                 China       0.00      0.00      0.00         2
               Croatia       0.00      0.00      0.00        15
        Czech Republic       0.00      0.00      0.00         2
               England       0.00      0.00      0.00        22
                France       0.82      0.79      0.80      5277
               Georgia       0.00      

### didnt wait--Gaussian Naive Bayes

### Categorical Bigram Naive Bayes

In [31]:
%%time

bigram_naive_bayes = make_pipeline(
    CountVectorizer(
        stop_words='english',
        binary=True,
        ngram_range=(1, 2)
    ),
    MultinomialNB()
)

bigram_naive_bayes.fit(X_train, y_train)

print(f'Accuracy: {bigram_naive_bayes.score(X_test, y_test)} \n')
print(classification_report(y_test, bigram_naive_bayes.predict(X_test)))

Accuracy: 0.7947503799836315 



/Users/Eric/anaconda3/lib/python3.7/site-packages/sklearn/metrics/classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


                        precision    recall  f1-score   support

             Argentina       0.97      0.08      0.14       954
               Armenia       0.00      0.00      0.00         3
             Australia       1.00      0.05      0.09       610
               Austria       0.94      0.27      0.42       760
Bosnia and Herzegovina       0.00      0.00      0.00         2
                Brazil       0.00      0.00      0.00        12
              Bulgaria       0.00      0.00      0.00        44
                Canada       0.00      0.00      0.00        63
                 Chile       0.89      0.20      0.33       996
                 China       0.00      0.00      0.00         2
               Croatia       0.00      0.00      0.00        15
        Czech Republic       0.00      0.00      0.00         2
               England       0.00      0.00      0.00        22
                France       0.75      0.90      0.82      5277
               Georgia       0.00      

### Word Count Logistic Regression

In [30]:
%%time

normalized_lr = make_pipeline(
    CountVectorizer(stop_words='english'),
    Normalizer(),
    LogisticRegression()
)

normalized_lr.fit(X_train, y_train)

print(f'Accuracy: {normalized_lr.score(X_test, y_test)} \n')
print(classification_report(y_test, normalized_lr.predict(X_test)))

Accuracy: 0.8462527768034608 

                        precision    recall  f1-score   support

             Argentina       0.75      0.39      0.51       954
               Armenia       0.00      0.00      0.00         3
             Australia       0.74      0.48      0.59       610
               Austria       0.80      0.60      0.69       760
Bosnia and Herzegovina       0.00      0.00      0.00         2
                Brazil       0.00      0.00      0.00        12
              Bulgaria       0.71      0.11      0.20        44
                Canada       0.00      0.00      0.00        63
                 Chile       0.64      0.52      0.57       996
                 China       0.00      0.00      0.00         2
               Croatia       0.00      0.00      0.00        15
        Czech Republic       0.00      0.00      0.00         2
               England       0.00      0.00      0.00        22
                France       0.79      0.88      0.83      5277
        

### TF-IDF Logistic Regression Model

In [29]:
%%time

# TF-IDF Logistic Regression Model

tfidf_logistic_regression = make_pipeline(
   CountVectorizer(strip_accents='ascii',
                   stop_words='english'),
   TfidfTransformer(),
   LogisticRegression(),
   
   
)

tfidf_logistic_regression.fit(X_train, y_train)

print(f'Accuracy: {tfidf_logistic_regression.score(X_test, y_test)} \n')
print(classification_report(y_test, tfidf_logistic_regression.predict(X_test)))

Accuracy: 0.8575353677072373 

                        precision    recall  f1-score   support

             Argentina       0.75      0.42      0.54       954
               Armenia       0.00      0.00      0.00         3
             Australia       0.78      0.52      0.62       610
               Austria       0.82      0.64      0.72       760
Bosnia and Herzegovina       0.00      0.00      0.00         2
                Brazil       0.00      0.00      0.00        12
              Bulgaria       0.77      0.23      0.35        44
                Canada       0.00      0.00      0.00        63
                 Chile       0.62      0.54      0.58       996
                 China       0.00      0.00      0.00         2
               Croatia       0.00      0.00      0.00        15
        Czech Republic       0.00      0.00      0.00         2
               England       0.00      0.00      0.00        22
                France       0.81      0.89      0.84      5277
        

### Normalized Word Count Logistic Regression

In [28]:
%%time

normalized_lr = make_pipeline(
    CountVectorizer(stop_words='english'),
    Normalizer(),
    LogisticRegression()
)

normalized_lr.fit(X_train, y_train)

print(f'Accuracy: {normalized_lr.score(X_test, y_test)} \n')
print(classification_report(y_test, normalized_lr.predict(X_test)))

Accuracy: 0.8462527768034608 

                        precision    recall  f1-score   support

             Argentina       0.75      0.39      0.51       954
               Armenia       0.00      0.00      0.00         3
             Australia       0.74      0.48      0.59       610
               Austria       0.80      0.60      0.69       760
Bosnia and Herzegovina       0.00      0.00      0.00         2
                Brazil       0.00      0.00      0.00        12
              Bulgaria       0.71      0.11      0.20        44
                Canada       0.00      0.00      0.00        63
                 Chile       0.64      0.52      0.57       996
                 China       0.00      0.00      0.00         2
               Croatia       0.00      0.00      0.00        15
        Czech Republic       0.00      0.00      0.00         2
               England       0.00      0.00      0.00        22
                France       0.79      0.88      0.83      5277
        

### Bigram TF-IDF Logistic Regression

In [27]:
%%time

bigram_tfidf_logistic_regression = make_pipeline(
    CountVectorizer(
        stop_words='english',
        ngram_range=(1,2)
    ),
    TfidfTransformer(),
    LogisticRegression()
)

bigram_tfidf_logistic_regression.fit(X_train, y_train)

print(f'Accuracy: {bigram_tfidf_logistic_regression.score(X_test, y_test)} \n')
print(classification_report(y_test, bigram_tfidf_logistic_regression.predict(X_test)))

Accuracy: 0.8539985969835145 

                        precision    recall  f1-score   support

             Argentina       0.82      0.38      0.51       954
               Armenia       0.00      0.00      0.00         3
             Australia       0.82      0.44      0.58       610
               Austria       0.86      0.59      0.70       760
Bosnia and Herzegovina       0.00      0.00      0.00         2
                Brazil       0.00      0.00      0.00        12
              Bulgaria       0.75      0.07      0.12        44
                Canada       0.00      0.00      0.00        63
                 Chile       0.68      0.52      0.59       996
                 China       0.00      0.00      0.00         2
               Croatia       0.00      0.00      0.00        15
        Czech Republic       0.00      0.00      0.00         2
               England       0.00      0.00      0.00        22
                France       0.80      0.90      0.85      5277
        

###  TF-IDF Logistic Regression with Automatic Corpus Specific Stop Words

In [26]:
%%time

mindf_max_tfidf_lr = make_pipeline(
    CountVectorizer(
        stop_words='english',
        min_df=2,
        max_df=.9
    ),
    TfidfTransformer(),
    LogisticRegression()
)

mindf_max_tfidf_lr.fit(X_train, y_train)

print(f'Accuracy: {mindf_max_tfidf_lr.score(X_test, y_test)} \n')
print(classification_report(y_test, mindf_max_tfidf_lr.predict(X_test)))



/Users/Eric/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/Eric/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


Accuracy: 0.8578861218285981 

                        precision    recall  f1-score   support

             Argentina       0.75      0.42      0.54       954
               Armenia       0.00      0.00      0.00         3
             Australia       0.78      0.53      0.63       610
               Austria       0.82      0.64      0.72       760
Bosnia and Herzegovina       0.00      0.00      0.00         2
                Brazil       0.00      0.00      0.00        12
              Bulgaria       0.77      0.23      0.35        44
                Canada       0.00      0.00      0.00        63
                 Chile       0.63      0.54      0.58       996
                 China       0.00      0.00      0.00         2
               Croatia       0.00      0.00      0.00        15
        Czech Republic       0.00      0.00      0.00         2
               England       0.00      0.00      0.00        22
                France       0.81      0.89      0.84      5277
        

### caused reboot-Linear Discrete Analysis

### TF-IDF Linear SVC

In [38]:
%%time

tfidf_svc = make_pipeline(
    CountVectorizer(
        stop_words='english',
        #min_df = 2,      # ed
        #max_df = .2      # ed
    ),
    TfidfTransformer(),
    #DenseTransformer(),
    LinearSVC()
)

tfidf_svc.fit(X_train, y_train)

print(f'Accuracy: {tfidf_svc.score(X_test, y_test)} \n')
print(classification_report(y_test, tfidf_svc.predict(X_test)))

Accuracy: 0.881942008651935 

                        precision    recall  f1-score   support

             Argentina       0.65      0.51      0.57       954
               Armenia       1.00      0.67      0.80         3
             Australia       0.80      0.66      0.72       610
               Austria       0.81      0.73      0.77       760
Bosnia and Herzegovina       0.00      0.00      0.00         2
                Brazil       1.00      0.08      0.15        12
              Bulgaria       0.83      0.57      0.68        44
                Canada       0.67      0.13      0.21        63
                 Chile       0.58      0.58      0.58       996
                 China       1.00      1.00      1.00         2
               Croatia       1.00      0.27      0.42        15
        Czech Republic       0.00      0.00      0.00         2
               England       0.79      0.50      0.61        22
                France       0.83      0.90      0.86      5277
         

### warnings--dnf   TF-IDF Logistic Lasso

### coef do not converge -- TF-IDF Logistic Ridge

### error-TF-IDF Logistic Elasticnet

### Looking at Top 20 Countries with top 3 accurate Models from above full data set work

In [39]:
# use Counter to get the top 20 wine countries

counter = Counter(data['country'].tolist())
top_20_countries = {i[0]: idx for idx, i in enumerate(counter.most_common(20))}
top20_countries_data = data[data['country'].map(lambda x: x in top_20_countries)]

top20_countries_data['country'].value_counts()

US              70956
Italy           31757
France          26183
Spain            8609
Portugal         6913
Chile            5284
Argentina        4567
Austria          3866
Australia        3148
Germany          2917
New Zealand      1935
South Africa     1582
Israel            689
Greece            566
Canada            336
Uruguay           228
Bulgaria          202
Hungary           187
Romania           143
England           123
Name: country, dtype: int64

In [41]:
# filter out all extraneaous columns

df = top20_countries_data.filter(["country", "description"], axis = 1)
df.head()

country                                        description
0  Portugal  This is ripe and fruity, a wine that is smooth...
1        US  Tart and snappy, the flavors of lime flesh and...
2        US  Pineapple rind, lemon pith and orange blossom ...
3        US  Much like the regular bottling from 2012, this...
4     Spain  Blackberry and raspberry aromas show a typical...

In [43]:
df.count()

country        170191
description    170191
dtype: int64

In [45]:
# split the dataset into training and testing datasets

X=df['description']
y=df['country']

X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=42)

### TF-IDF Logistic Regression Model:  Pvs Accuracy: 0.8575353677072373  <<<<<<<<  #3


In [46]:
%%time

# TF-IDF Logistic Regression Model

tfidf_logistic_regression = make_pipeline(
   CountVectorizer(strip_accents='ascii',
                   stop_words='english'),
   TfidfTransformer(),
   LogisticRegression(),
      
)

tfidf_logistic_regression.fit(X_train, y_train)

print(f'Accuracy: {tfidf_logistic_regression.score(X_test, y_test)} \n')
print(classification_report(y_test, tfidf_logistic_regression.predict(X_test)))

/Users/Eric/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/Eric/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


Accuracy: 0.8621581127530186 

              precision    recall  f1-score   support

   Argentina       0.68      0.41      0.51       893
   Australia       0.74      0.47      0.58       606
     Austria       0.82      0.62      0.71       803
    Bulgaria       0.90      0.21      0.35        42
      Canada       0.00      0.00      0.00        76
       Chile       0.72      0.54      0.62      1092
     England       0.00      0.00      0.00        29
      France       0.80      0.88      0.84      5155
     Germany       0.77      0.66      0.71       538
      Greece       0.74      0.36      0.49       116
     Hungary       1.00      0.15      0.27        39
      Israel       0.91      0.46      0.61       128
       Italy       0.98      0.96      0.97      6477
 New Zealand       0.86      0.42      0.56       398
    Portugal       0.94      0.50      0.66      1352
     Romania       1.00      0.09      0.17        22
South Africa       0.76      0.50      0.61       

###  TF-IDF Logistic Regression with Automatic Corpus Specific Stop Words:  Pvs Accuracy: 0.8578861218285981   <<<<<<<  #2   

In [47]:
%%time

mindf_max_tfidf_lr = make_pipeline(
    CountVectorizer(
        stop_words='english',
        min_df=2,
        max_df=.9
    ),
    TfidfTransformer(),
    LogisticRegression()
)

mindf_max_tfidf_lr.fit(X_train, y_train)

print(f'Accuracy: {mindf_max_tfidf_lr.score(X_test, y_test)} \n')
print(classification_report(y_test, mindf_max_tfidf_lr.predict(X_test)))


Accuracy: 0.8619524662886688 

              precision    recall  f1-score   support

   Argentina       0.68      0.41      0.51       893
   Australia       0.75      0.47      0.58       606
     Austria       0.82      0.61      0.70       803
    Bulgaria       0.90      0.21      0.35        42
      Canada       0.00      0.00      0.00        76
       Chile       0.72      0.54      0.62      1092
     England       0.00      0.00      0.00        29
      France       0.80      0.88      0.84      5155
     Germany       0.78      0.67      0.72       538
      Greece       0.76      0.38      0.51       116
     Hungary       1.00      0.15      0.27        39
      Israel       0.91      0.46      0.61       128
       Italy       0.98      0.96      0.97      6477
 New Zealand       0.83      0.41      0.55       398
    Portugal       0.94      0.50      0.66      1352
     Romania       1.00      0.09      0.17        22
South Africa       0.75      0.50      0.60       

### TF-IDF Linear SVC:  Pvs  Accuracy: 0.881942008651935     <<<<<   #1


In [48]:
%%time

tfidf_svc = make_pipeline(
    CountVectorizer(
        stop_words='english',
        #min_df = 2,      # ed
        #max_df = .2      # ed
    ),
    TfidfTransformer(),
    #DenseTransformer(),
    LinearSVC()
)

tfidf_svc.fit(X_train, y_train)

print(f'Accuracy: {tfidf_svc.score(X_test, y_test)} \n')
print(classification_report(y_test, tfidf_svc.predict(X_test)))

Accuracy: 0.8863362613472782 

              precision    recall  f1-score   support

   Argentina       0.64      0.53      0.58       893
   Australia       0.78      0.63      0.69       606
     Austria       0.84      0.71      0.77       803
    Bulgaria       0.89      0.60      0.71        42
      Canada       0.71      0.07      0.12        76
       Chile       0.69      0.60      0.64      1092
     England       0.90      0.62      0.73        29
      France       0.83      0.89      0.86      5155
     Germany       0.77      0.77      0.77       538
      Greece       0.79      0.72      0.76       116
     Hungary       1.00      0.49      0.66        39
      Israel       0.87      0.70      0.78       128
       Italy       0.98      0.98      0.98      6477
 New Zealand       0.80      0.61      0.69       398
    Portugal       0.89      0.59      0.71      1352
     Romania       1.00      0.32      0.48        22
South Africa       0.75      0.66      0.70       